### PR TITLE
highly variable gene annotation

### PR DESCRIPTION
**Pull request by bkmartinjr**
Add SOMA-optimized equivalent of scanpy.pp.highly_variable_genes, flavor="seurat_v3"

Example:

```
In [1]: import cellxgene_census
   ...: from cellxgene_census.experimental.pp import get_highly_variable_genes
   ...: 
   ...: with cellxgene_census.open_soma(census_version="stable") as census:
   ...:     hvg = get_highly_variable_genes(
   ...:         census,
   ...:         organism="Mus musculus",
   ...:         obs_value_filter="is_primary_data == True and tissue_general == 'lung'",
   ...:         n_top_genes = 500
   ...:     )
   ...: 
   ...: hvg[hvg.highly_variable]
   ...: 
The "stable" release is currently 2023-05-15. Specify 'census_version="2023-05-15"' in future calls to open_soma() to ensure data consistency.
Out[1]: 
                means    variances  highly_variable_rank  variances_norm  highly_variable
soma_joinid                                                                              
85           0.104754    93.239460                 343.0        3.671225             True
102          0.772019  2061.330152                 332.0        3.688259             True
115          0.103565    94.853489                 167.0        4.244059             True
214          0.048323    32.914252                 281.0        3.795256             True
336          0.010261     1.912341                 360.0        3.641506             True
...               ...          ...                   ...             ...              ...
29230        0.067882    47.751253                 133.0        4.462335             True
35183        0.019692     6.483923                 471.0        3.448803             True
46714        0.033726    27.929340                 423.0        3.517277             True
48253        0.006745     0.804313                 275.0        3.805323             True
49751        0.064489    45.016323                 497.0        3.421169             True

[500 rows x 5 columns]

In [2]: var_df = census['census_data']['mus_musculus'].ms['RNA'].var.read().concat().to_pandas().set_index('soma_joinid')

In [3]: var_df
Out[3]: 
                     feature_id                       feature_name  feature_length
soma_joinid                                                                       
0            ENSMUSG00000109644                      0610005C13Rik            3583
1            ENSMUSG00000108652                      0610006L08Rik            2128
2            ENSMUSG00000007777                      0610009B22Rik             998
3            ENSMUSG00000086714                      0610009E02Rik            1803
4            ENSMUSG00000043644                      0610009L18Rik             619
...                         ...                                ...             ...
52387        ENSMUSG00000081591                           Btf3-ps9             496
52388        ENSMUSG00000118710  mmu-mir-467a-3_ENSMUSG00000118710              83
52389        ENSMUSG00000119584                              Rn18s            1849
52390        ENSMUSG00000118538                            Gm18218             970
52391        ENSMUSG00000084217                           Setd9-ps             670

[52392 rows x 3 columns]

In [4]: combined_df = pd.concat([var_df, hvg], axis=1)

In [5]: combined_df[combined_df.highly_variable]
Out[5]: 
                     feature_id   feature_name  feature_length     means    variances  highly_variable_rank  variances_norm  highly_variable
soma_joinid                                                                                                                                 
85           ENSMUSG00000075511  1700001L05Rik            6112  0.104754    93.239460                 343.0        3.671225             True
102          ENSMUSG00000026831  1700007K13Rik             965  0.772019  2061.330152                 332.0        3.688259             True
115          ENSMUSG00000023873  1700010I14Rik            2292  0.103565    94.853489                 167.0        4.244059             True
214          ENSMUSG00000044916  1700029I15Rik            1348  0.048323    32.914252                 281.0        3.795256             True
336          ENSMUSG00000097675  1700101I11Rik            1868  0.010261     1.912341                 360.0        3.641506             True
...                         ...            ...             ...       ...          ...                   ...             ...              ...
29230        ENSMUSG00000096323        Gm20767            1427  0.067882    47.751253                 133.0        4.462335             True
35183        ENSMUSG00000051537         Gm5124            2742  0.019692     6.483923                 471.0        3.448803             True
46714        ENSMUSG00000092300           Cdk3            1548  0.033726    27.929340                 423.0        3.517277             True
48253        ENSMUSG00000105428        Mir3068              79  0.006745     0.804313                 275.0        3.805323             True
49751        ENSMUSG00000023795       Pisd-ps2            2108  0.064489    45.016323                 497.0        3.421169             True

[500 rows x 8 columns]
```

### CODE REVIEW

File: api/python/cellxgene_census/tests/experimental/pp/test_hvg.py
Comment (atolopko-czi) [inline api/python/cellxgene_census/tests/experimental/pp/test_hvg.py:35]: nit: move the cheap test to the front of the array

File: api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py
Comment (atolopko-czi) [inline api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py:254]: Would it be worth having this convenience wrapper return the `combined_df`, per the example in the PR summary?

File: api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py
Comment (atolopko-czi) [inline api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py:184]: nit: better name to differentiate it from the convenience wrapper method, below? `compute_highly_variable_genes`? I don't have any great naming suggestions, so feel free to leave as is.

**Review by atolopko-czi — APPROVED** [review]
Made some minor suggestions and nits, but can approve as is. I did _not_ review this deeply at an algorithmic level, though I did peruse all of the code.

Worth adding a notebook for this code?

Should `experimental/pp/__init__.py` include a comment to spell out that "pp" is an abbreviation for "preprocessing"?  I understand that name is coming from scanpy, so if it's assumed knowledge, no change needed.

File: api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_eager_iter.py
Comment (atolopko-czi) [inline api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_eager_iter.py:9]: This could be useful to have in a sibling package, since it's useful elsewhere and not preprocessing-specific. E.g., I suspect it can be leveraged by `pytorch.py`, I can move when it's incorporated elsewhere.

File: api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py
Comment (atarashansky) [inline api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py:124]: I think this is a good idea. I retry manually with jitter every time I get a loess error, anyway, so it seems reasonable to automate that.

File: api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_eager_iter.py
Comment (bkmartinjr) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_eager_iter.py:43]: Yes, but left in because I am consider it as a means of further performance enhancement.

File: api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py
Comment (bkmartinjr) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py:254]: It could, but that would require the fetch of the entire var dataframe, and I did not want to presume that the user would want that extra overhead. 

Most use cases I know of do _not_ fetch the entire var.

I'll add an example to the docstrings to show that use case.

File: api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_eager_iter.py
Comment (bkmartinjr) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_eager_iter.py:9]: feel free to grab/move it whenever you find a second use case. I didn't want to presume it would be useful until it was :-)

File: api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py
Comment (atolopko-czi) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py:303]: Not really, other than using a URL shortener, which seems like the wrong to do in a docstring.

File: api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py
Comment (bkmartinjr) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py:124]: OK, will do - it seems simple and very useful.

My feeling is that the parameterization of this should be a new argument called "max_jitter", which can be set to zero to get the old behavior.  The code would retry adding jitter until either success or max_jitter is exceeded.  I'll ping you for another review when it is added.

Comment (bkmartinjr) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py:124]: @atarashansky - any thoughts on what would be a reasonable range of jitter values to try before bailing?  It is a 64 bit float, so it can start small.

I'm going to start with  
> [  10 ** -p for p in range(18, 8, -1) ]

i.e.,

> [1e-18, 1e-17, 1e-16, 1e-15, 1e-14, 1e-13, 1e-12, 1e-11, 1e-10, 1e-09]

**Comment by bkmartinjr** [comment]
@atolopko-czi - I believe I addressed your comments, so feel free to take another look.  I don't plan on landing this until next week after Pablo has reviewed.

**Comment by atolopko-czi** [comment]
> @atolopko-czi - I believe I addressed your comments, so feel free to take another look. I don't plan on landing this until next week after Pablo has reviewed.

LGTM!

**Review by pablo-gar — APPROVED** [review]
I don't see anything obviously wrong and the test with the scanpy comparison is solid!!

**Comment by bkmartinjr** [comment]
@atarashansky - any last comments on the `loess` adaptive retry before I land this?

File: api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py
Comment (atarashansky) [inline api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py:60]: Should we allow users to specify `jitter` explicitly? If specified, we don't do the search across orders of magnitude and let the function fail with `ValueError` if there is a loess error.

I worry that the function may take a long time if the following conditions were met:
 - Loess errors are common
 - `1e-18` starting value for jitter is not enough to resolve loess

In your experimenting, what are the chances of the above conditions being met? If likely, then a `jitter` parameter would let users set a reasonably high value that is likely to work so they don't have to wait for the function to scan.

Otherwise, approving now as the current implementation looks good!

**Review by atarashansky — APPROVED** [review]

File: api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py
Comment (bkmartinjr) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py:60]: The amount of jitter needed is going to be very dataset dependent, so I think there are edge cases for almost any default value.  I picked a conservative range of values because I thought getting the minimum jitter that "solves" was more important than performance (performance being dominated by I/O - read - time in almost all cases).

The current API allows the user to control only the max jitter.  We could expand it to allow the user to specify the entire range (and step) to try. But that still leaves the issue of what a good default is, which will be where most users start.

@pablo-gar - appreciate your thoughts.

**Pull request merged**
